docs/marketplace/tutorials/fully-functioning-rag-application.md

```diff
@@ -0,0 +1,172 @@
+
+The prerequisites for this tutorial are as follows:
+
+- A connection string to a [MongoDB Atlas instance](https://www.mongodb.com/products/platform/atlas-database) (unfortunately, MongoDB on-premise installations do not currently support the retrieval of embeddings from documents).
+- An API key to communicate with [OpenAI](https://openai.com/).
+
+## How it works
+
+RAG stands for Retrieval-Augmented Generation, and it is a process to enhance the knowledge of a LLM (large language model), which is an algorithm capable of generating text in natural language. The goal of RAG is to create applications that can communicate with users and answer questions related to specific topics.
+
+To understand how RAG works, let's consider an example. Imagine you want to create a chatbot that can answer questions about the internal documentation of the product your company sells. This documentation can be extensive, consisting of multiple files, just like this documentation. To handle this large amount of text, it can be divided into smaller pieces called "chunks". These chunks can be split in various ways, such as by paragraph, by page, or by the semantic meaning of the text. Each chunk can then be transformed into a single document and stored in a database.
+
+This will allow us to compare the question asked by a user to the text saved in the database. In order to efficiently retrieve documents from the database, we need to find a reliable method to compare the meaning of the question with the meaning of the chunks we want to use. To achieve this, we can create embeddings of our chunks. The text is given to a mathematical model that will generate a multidimensional vector composed of millions of values between 0 and 1, effectively representing the meaning of the text.
+
+By doing this, whenever our chatbot receives a question, we can:
+
+- Generate embeddings of our question.
+- Search the database for documents with embeddings that are similar in meaning to the embeddings of the question.
+- Retrieve the matched documents and use their text to ensure that the chatbot provides an appropriate answer.
+
+OpenAI has several [embedding models](https://platform.openai.com/docs/guides/embeddings/embedding-models), and MongoDB proposes the [Vector Search Index](https://www.mongodb.com/docs/atlas/atlas-vector-search/vector-search-overview/) to efficiently search in a database for documents where embeddings are similar in meaning to our question.
+The embeddings from the internal documentation can be created only once, or whenever there is an important update. There are several methods to execute this process. If the documentation is available on a website, the most common and straightforward approach is to perform [web scraping](https://en.wikipedia.org/wiki/Web_scraping). This involves downloading all the pages of the website, removing any unnecessary information such as HTML tags, headers, footers, and styles, and generating embeddings from the remaining text.
```

Comment:
_> The embeddings from the internal documentation can be created only once, or whenever there is an important update._

What do you mean by "they can be created only once?", and what do you refer to as "internal documentation"?

Reply:
_they can be created only once_, meaning could be a one-time process, you don't need to create them every time you use the chatbot or every time you run the service. You might re-create them once the documentation (the original knowledge base) changes. I should rephrase this.

_Internal documentation_ as stated in line 25, it's an example to simplify references to the texts that will be used for the RAG.